griptape/tasks/base_rule_task.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import annotations
+
+from abc import ABC
+from typing import Any
+
+from attr import define, field
+
+from griptape.rules import Ruleset, Rule
+from griptape.tasks import BaseTask
+
+
+@define
+class BaseRuleTask(BaseTask, ABC):
+    DEFAULT_RULESET_NAME = "Default Ruleset"
+    ADDITIONAL_RULESET_NAME = "Additional Ruleset"
+
+    rulesets: list[Ruleset] = field(factory=list, kw_only=True)
+    rules: list[Rule] = field(factory=list, kw_only=True)
+    context: dict[str, Any] = field(factory=dict, kw_only=True)
+
+    @rulesets.validator  # pyright: ignore
+    def validate_rulesets(self, _, rulesets: list[Ruleset]) -> None:
+        if not rulesets:
+            return
+
+        if self.rules:
+            raise ValueError("Can't have both rulesets and rules specified.")
+
+    @rules.validator  # pyright: ignore
+    def validate_rules(self, _, rules: list[Rule]) -> None:
+        if not rules:
+            return
+
+        if self.rulesets:
+            raise ValueError("Can't have both rules and rulesets specified.")
+
+    @property
+    def all_rulesets(self) -> list[Ruleset]:
+        structure_rulesets = []
+
+        if self.structure:
+            if self.structure.rulesets:
+                structure_rulesets = self.structure.rulesets
+            elif self.structure.rules:
+                structure_rulesets = [Ruleset(name=self.DEFAULT_RULESET_NAME, rules=self.structure.rules)]
+
+        task_rulesets = []
+        if self.rulesets:
+            task_rulesets = self.rulesets
+        elif self.rules:
+            if structure_rulesets:
+                task_ruleset_name = self.ADDITIONAL_RULESET_NAME
+            else:
+                task_ruleset_name = self.DEFAULT_RULESET_NAME
+
+            task_rulesets = [Ruleset(name=task_ruleset_name, rules=self.rules)]
+
+        return structure_rulesets + task_rulesets
+
+    @property
+    def full_context(self) -> dict[str, Any]:
+        if self.structure:
+            structure_context = self.structure.context(self)
+
+            structure_context.update(self.context)
+
+            return structure_context
+        else:
+            return {}
```

Comment:
Context-related code is relevant to image generation tasks as well, so I've moved it to BaseRuleTask. This dulls the precision of the naming a bit, but it could be overkill to introduce a BaseContextTask. Maybe context code should be promoted to BaseTask?